Add ContainerRuntimeInterface.

yifan-gu opened the pull request:
Rename DockerInterface to ContainerRuntimeInterface.
Rename dockerClient field in Kubelet to containerRuntime.
Rename DockerClient type in KubeletConfig to ContainerRuntime.

As the first step for #4404, here I pulled DockerInterface type and dockerClient out from kubelet.
Next I plan to modify the functions in ContainerRuntimeInterface to make them non-docker-specified.

PTAL. Thanks!

Commit:
Add ContainerRuntimeInterface.

Rename DockerInterface to ContainerRuntimeInterface.
Rename dockerClient field in Kubelet to containerRuntime.
Rename DockerClient type in KubeletConfig to ContainerRuntime.

thockin commented on pkg/kubelet/cadvisor.go:
```diff
@@ -59,7 +59,7 @@ func (kl *Kubelet) GetContainerInfo(podFullName string, uid types.UID, container
 		return nil, nil
 	}
-	dockerContainers, err := dockertools.GetKubeletDockerContainers(kl.dockerClient, false)
+	dockerContainers, err := dockertools.GetKubeletDockerContainers(kl.containerRuntime, false)
```
Design question - do we expect that kubelet supports a single container runtime or that it supports rocket and docker at the same time on the same machine?  If the former, how will we spec which one to run?  If the latter, do we support them by having kl.dockerClient and kl.rocketRuntime as 2 fields or do we hide them behind a runtime mux?

yifan-gu replied:
@thockin IMO, since docker and rocket are going to support both docker image and aci. I think one container runtime is enough, and I think we can spec which one to run in the configure :)

derekwaynecarr replied:
Does every kubelet in a kubernetes cluster use the same container 

Sent from my iPhone

> On Feb 13, 2015, at 5:58 PM, Yifan Gu <notifications@github.com> wrote:
> 
> In pkg/kubelet/cadvisor.go:
> 
> > @@ -59,7 +59,7 @@ func (kl *Kubelet) GetContainerInfo(podFullName string, uid types.UID, container
> >  	if cc == nil {
> >  		return nil, nil
> >  	}
> > -	dockerContainers, err := dockertools.GetKubeletDockerContainers(kl.dockerClient, false)
> > +	dockerContainers, err := dockertools.GetKubeletDockerContainers(kl.containerRuntime, false)
> @thockin IMO, since docker and rocket are going to support both docker image and aci. I think one container runtime is enough, and I think we can spec which one to run in the configure :)
> 
> —
> Reply to this email directly or view it on GitHub.
>

derekwaynecarr replied:
Does every kubelet in a k8s cluster use the same runtime?

Sent from my iPhone

> On Feb 13, 2015, at 5:58 PM, Yifan Gu <notifications@github.com> wrote:
> 
> In pkg/kubelet/cadvisor.go:
> 
> > @@ -59,7 +59,7 @@ func (kl *Kubelet) GetContainerInfo(podFullName string, uid types.UID, container
> >  	if cc == nil {
> >  		return nil, nil
> >  	}
> > -	dockerContainers, err := dockertools.GetKubeletDockerContainers(kl.dockerClient, false)
> > +	dockerContainers, err := dockertools.GetKubeletDockerContainers(kl.containerRuntime, false)
> @thockin IMO, since docker and rocket are going to support both docker image and aci. I think one container runtime is enough, and I think we can spec which one to run in the configure :)
> 
> —
> Reply to this email directly or view it on GitHub.
>

yifan-gu replied:
@derekwaynecarr They don't have to as we can provide different cmd options to different kubelet. But I am not seeing reasons why we need different runtime in the same cluster. Maybe for live updating the cluster?

bgrant0607 replied:
For now, I think it's fine to assume that all Kubelets are configured with the same runtime. Otherwise, we'd need to surface information about which runtime was supported by each node so that the scheduler could choose nodes able to run the images specified by the pods' containers.

thockin replied:
It doesn't look to me that docker is going to support ACI.  Does that mean that we need to revisit this?  Or is it sufficient to say that we'll inject a mux here and that the Pod spec has enough information to route to the appropriate concrete implementation?

Commit:
Add DockerRuntime.

Wrap DockerInterface with DockerRuntime to implement ContainerRuntimeInterface.

bgrant0607 commented on pkg/kubelet/container/runtime.go:
```diff
@@ -0,0 +1,39 @@
+
+// ContainerRuntimeInterface is the abstract interface for container runtime.
+type ContainerRuntimeInterface interface {
+	ListContainers(options docker.ListContainersOptions) ([]docker.APIContainers, error)
```
There are a number of docker-specific arguments here. Any ideas about what to do about that?

yifan-gu replied:
How about creating something like ListContainerOptions, which we can add fields as we need.

bgrant0607 replied:
Almost every call contains a parameter from go-dockerclient: ListContainerOptions, Container, CreateContainerOptions, ... Really those parameter structs have to be part of the interface. 

I'm fine with this PR as a first step, but you should ask fsouza whether he'd accept a major refactoring of the library to target multiple container runtimes at that level. Otherwise, we should consider forking that library. 

For now, a simple change would be to use an alias for that package here that is more generic.

yifan-gu commented:
@bgrant0607 As I am changing the ListContainers's return type. Here are the options just came to my mind.

1): Create a ```Container``` type under ```pkg/kubelet/container/``` with fields similar to ```docker.APIContainer```. But this seems to be repeating as we already have ```api.Container```, and might cause some confusion to users.

2): Use ```api.Container```. However the ```api.Container``` doesn't have fields like ```ID```, our options here might be:
      a) Modify the ```api.Container```, add ```ID``` field in ```api.Container``` (which cause small API changes)
      b) Use container's Name as ```ID```, since "each container in a pod must a unique name", and it also seems to be unique on the same machine.

thockin commented:
Please do not overload meanings of api.* fields.

On Mon, Feb 16, 2015 at 5:31 PM, Yifan Gu <notifications@github.com> wrote:

> @bgrant0607 <https://github.com/bgrant0607> As I am changing the
> ListContainers's return type. Here are the options just came to my mind.
>
> 1): Create a Container type under pkg/kubelet/container/ with fields
> similar to docker.APIContainer. But this seems to be repeating as we
> already have api.Container, and might cause some confusion to users.
>
> 2): Use api.Container. However the api.Container doesn't have fields like
> ID, our options here might be:
> a) Modify the api.Container, add ID field in api.Container (which cause
> small API changes)
> b) Use container's Name as ID, since "each container in a pod must a
> unique name", and it also seems to be unique on the same machine.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4431#issuecomment-74603865>
> .
>

bgrant0607 commented:
+1 to what @thockin wrote: Let's not conflate the container runtime layer with the Kubernetes api at this point. I'd love to make the Kubernetes API fully composable with the underlying container runtime layer rather than wrap it, but the container API is currently not adequate for our needs and also provides flexibility that we can't expose without compromising correctness, security, privacy, and/or isolation.

yifan-gu commented:
@thockin @bgrant0607 Agree. So I just added pkg/kubelet/container.Container, which is same as docker.APIContainer for now. After all these clean ups we can revise this struct as we add rocket support.

Commit:
Change docker.APIContainers to container.Container.

Also change the parameter and return types of container.ListContainers().

Commit:
Refactor pkg/kubelet/container.ContainerRuntimeInterface.InspectContainer.

Change the return type from *docker.Container to *container.Container.

Commit:
Refactor pkg/kubelet/container.ContainerRuntimeInterface.ListContainers.

Change the return type from []Container to []*Containers.

Commit:
Refactor pkg/kubelet/container.ContainerRuntimeInterface.CreateContainer.

Change docker.CreateContainerOptions to container.CreateContainerOptions.
Change the return type from *docker.Container to *container.Container.

Commit:
Refactor pkg/kubelet/container.ContainerRuntimeInterface.StartContainer.

Change docker.HostConfig to container.HostConfig, eliminate not used fields.

Commit:
Refactor pkg/kubelet/container.ContainerRuntimeInterface.RemoveContainer.

Change docker.RemoveContainerOptions to container.RemoveContainerOptions.
Also add json description for container.ListContainersOptions.

Commit:
Refactor pkg/kubelet/container.ContainerRuntimeInterface.InspectImage.

Change return type from *docker.Image to *container.Image.

Commit:
Refactor pkg/kubelet/container.ContainerRuntimeInterface.ListImages

Change docker.ListImagesOptions to container.ListImageOptions.
Change return type from []docker.APIImage to []*container.Image.

Commit:
Rename container.ContainerRuntimeInterface to container.Runtime.

Also clean up DockerRuntime a little bit.

Commit:
Refactor pkg/kubelet/container.Runtime.PullImage.

Change docker.PullImageOptions to container.PullImageOptions.
Also embed docker.AuthConfiguration into container.PullImageOptions.

Commit:
Refactor pkg/kubelet/container.Runtime.Logs.

Change docker.LogsOptions to container.LogsOptions.

Commit:
Refactor pkg/kubelet/container.Runtime.Version.

Change return type from *docker.Env to []string.

Commit:
Refactor pkg/kubelet/container.Runtime.CreateExec.

Also pkg/kubelet/container.Runtime.StartExec.

yifan-gu commented:
@thockin @bgrant0607 I finished moving DockerInterface to container.Runtime Interface. 
PTAL. If lgty then I can squash the commits.

Next I will implement those interfaces with Rocket, probably also add/remove fields in some structs defined in types.go.

Thank you.

Commit:
Update comment.

Change ContainerRuntime to container.Runtime.

thockin commented on pkg/kubelet/container/types.go:
```diff
@@ -0,0 +1,173 @@
+)
+
+// Container represents a container.
+type Container struct {
```
I know this is "whatever docker does" for now, but can we get some comments on these fields?

thockin commented on pkg/kubelet/container/types.go:
```diff
@@ -0,0 +1,173 @@
+// Container represents a container.
+type Container struct {
+	ID              string            `json:"id"`
+	Name            string            `json:"name,omitempty"`
```
Do we need literally everything docker exposes, or can we slim it down and correct historical mistakes?  Like, do we need 2 name fields?

thockin commented on pkg/kubelet/container/types.go:
```diff
@@ -0,0 +1,173 @@
+}
+
+// PortBinding represents the host/container port mapping.
+type PortBinding struct {
```
Again, I know this is more or less direct from docker, but can we clean this up?  What is the meaning of Port vs PortBinding, for instance.

If we now own this abstraction layer, can we make it something that makes sense?

thockin replied:
As a thought experiment, if we commented out all of the fields in each of these structs and then tried to compile, which ones would we need to uncomment in order to make it build and run?  Do we need anything more than those?

thockin commented:
This is a large piece of work, so I will probably accept "we'll do that later" to many of my questions.

I am not final approver of this - @dchen1107

xiang90 commented on pkg/kubelet/container/types.go:
```diff
@@ -0,0 +1,173 @@
+	AttachStdin  bool     `json:"attachStdin,omitempty"`
+	AttachStdout bool     `json:"attachStdout,omitempty"`
+	AttachStderr bool     `json:"attachStderr,omitempty"`
+	Tty          bool     `json:"tty,omitempty"`
```
Tty -> TTY
https://github.com/golang/go/wiki/CodeReviewComments#initialisms

xiang90 commented on pkg/kubelet/container/types.go:
```diff
@@ -0,0 +1,173 @@
+type StartExecOptions struct {
+	Detach bool `json:"detach,omitempty"`
+
+	Tty bool `json:"tty,omitempty"`
```
TTY

yifan-gu commented:
@thockin Thanks for comments. Actually I removed a lot of unused stuffs, though there must be something left as you pointed out.
I am going to clean up this abstraction layer in the following up PRs as I implement these interfaces for rocket, cuz by that time I think we can be more clear about what we need and what we can omit. Does that sound good?

Will remove some obvious useless stuffs, add comments and rebase these two days.

Commit:
Cleanup types.

Remove Container.Names.
Also fixed some name convention.

Commit:
Refactor container.Container.

Remove Ports field.
Also refactor dockertools/docker.go, change var []*foo to make([]*foo).

Commit:
Add tests for docker.ConvertContainer and docker.ConvertAPIContainer.

Commit:
Added more comments to types.go

Also refactor container.Container to remove some not used fields.

yifan-gu commented:
@thockin @dchen1107 @bgrant0607 
I just added more comments and did some refactor to remove some unused fields.
PTAL, thanks!
cc/@xiang90

thockin commented:
I am still anxious about this overall, but won't block it.  The factorings in this new API are essentially random.  In many places it is not at all what we would do if we were not copying an existing API.

yifan-gu commented:
@thockin Yes, I understand. So this is just the temporal patch, the final goal is to bring rocket into the kubelet and by that time we can have a REAL abstraction layer.
Maybe we can do it this way:
I'm going to write the rocket integration code now, obviously some changes will be made to these APIs, so we can keep this PR open until it is stabilized.

bgrant0607 commented on pkg/kubelet/container/types.go:
```diff
@@ -0,0 +1,202 @@
+type Container struct {
+	// The id of the container.
+	ID string `json:"id"`
+	// The name of the container.
```
Isn't ID optional, in the case that a name is provided?

bgrant0607 replied:
Actually, do we need the json tags?

bgrant0607 commented on pkg/kubelet/container/types.go:
```diff
@@ -0,0 +1,202 @@
+	Env []string `json:"env,omitempty"`
+}
+
+// State represents the state of a container.
```
Except for removal of the yaml tags, this is a copy+paste of https://github.com/fsouza/go-dockerclient/blob/master/container.go#L89. 

I know I asked you to address the issue of Dockerisms in this API, but this PR doesn't do that. For now, probably it would be better to go back to using dockerclient types directly, until we figure out what API we really want, as per @thockin's questions (deleting unused fields, etc.). 

Future use cases we already know we'll likely want to support:
* Pure Java applications (Fabric8)
* Non-Linux platforms (MacOS, BSD, Windows)
* Legacy (pre-Docker) container runtimes and package formats

bgrant0607 commented:
@yifan-gu What do you plan to do next?

yifan-gu commented:
@bgrant0607 
I am now doing a docker-free prototype, also I changed `syncPod(pod, containers)` to `syncPod(pod, pod)`, which fits the rocket's model better IMO, I will PR it when OK, and then figure out how to combine docker with it.
You can close this PR for now, thank you!
